embedded-keycloak/src/main/java/com/playtika/test/keycloak/EmbeddedKeycloakDependenciesAutoConfiguration.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2018 Playtika
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.playtika.test.keycloak;
+
+import static com.playtika.test.keycloak.KeycloakProperties.BEAN_NAME_EMBEDDED_KEYCLOAK;
+
+import com.playtika.test.common.spring.DependsOnPostProcessor;
+import org.keycloak.adapters.springboot.KeycloakAutoConfiguration;
+import org.keycloak.adapters.springboot.KeycloakSpringBootConfigResolver;
+import org.springframework.beans.factory.config.BeanFactoryPostProcessor;
+import org.springframework.boot.autoconfigure.AutoConfigureAfter;
+import org.springframework.boot.autoconfigure.AutoConfigureOrder;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@AutoConfigureOrder
+@ConditionalOnClass(KeycloakAutoConfiguration.class)
```

Comment:
Does key-cloack has any official client? It is better to add conditional on the official client, because not all users will be using spring adapter

Reply:
That is a good question. It actually took me some time to understand what the purpose of these post processor dependencies are. As the Keycloak library itself is spring-less, I tried to find something meaningful in the Keycloak Boot starter, as without Boot people will not use `testcontainers-spring-boot`.

I am open for better candidates here! Please make suggestions.

Reply:
https://github.com/keycloak/keycloak/blob/master/integration/admin-client/src/main/java/org/keycloak/admin/client/Keycloak.java This one?)

Reply:
to be honest don't really understand why we need here ```org.keycloak.adapters.springboot``` at all
either create additional Autoconfiguration for users which want to use this integration(and add corresponding test, since by now tests contains pure http test, not including this integration lib anywhere) or remove this dependency at all...

Reply:
> As the Keycloak library itself is spring-less
Sorry, my first comment was not quite correct. As I understand, there are 2 approaches to integrate Keycloak: directly via spring-less client and via spring-adapter. So that means that some users may use spring-less Keycloak library. And currently `EmbeddedKeycloakDependenciesAutoConfiguration` setups dependencies only for the spring-adapter, but we do want to setup dependencies for the spring-less client as well.

> either create additional Autoconfiguration for users which want to use this integration(and add corresponding test, since by now tests contains pure http test, not including this integration lib anywhere) or remove this dependency at all...
Separate autoconfiguration may be more straightforward I think, but removing support completely will mean that official KeyCloak spring-adapter will have incorrect shutdown order (?) @holgerstolzenberg  am I right?

Reply:
Okay agreeded, I have to think about this a bit, but yes we should also provide the container if vanilla Keycloak is on the classpath.

The question is: If vanilla Keycloak is on the classpath, do we really need the `DependsOnPostProcessor` stuff then?

Reply:
It depends on whether vanilla Keycloak needs to be shutdown **before** keycloak container

Reply:
Hm, no matter if we use the Keycloak vanilla or with Spring, the `keycloak-core` dependency will always be on the classpath. But we cannot assume if ANY related spring bean is in the context as one could use keycloak just in a method without having a bean for it.

So in general for me it is:
`StartContainer => RunTests => ShutdownContainer`

So container always needs to come first and last. This makes it really hard to define the condition the DependsOnPostprocessor should be registered with…

Reply:
I think we have some misunderstanding here :)
I would like to have something similar to this:
```
@Configuration
@AutoConfigureOrder
@ConditionalOnClass(KeycloakSecurityContext.class)
@ConditionalOnProperty(name = "embedded.keycloak.enabled", matchIfMissing = true)
@AutoConfigureAfter(KeycloakAutoConfiguration.class)
public class EmbeddedKeycloakDependenciesAutoConfiguration {

// any other bean post processors here that are needed for vanilla keycloak client

@ConditionalOnClass(KeycloakAutoConfiguration.class)
public static class SpringAdapterDependenciesConfiguration {
    @Bean
    public BeanFactoryPostProcessor keycloakSpringBootConfigResolverDependencyPostProcessor() {
        return new DependsOnPostProcessor(KeycloakSpringBootConfigResolver.class,
            new String[]{BEAN_NAME_EMBEDDED_KEYCLOAK});
    }
}



}
```

Reply:
> // any other bean post processors here that are needed for vanilla keycloak client
For any vanilla Keycloak usage there won´t be any bean to register a post processor upon…